lib/internal/modules/esm/loader.js

```diff
@@ -514,9 +518,16 @@ function createModuleLoader(useCustomLoadersIfPresent = true) {
       !require('internal/modules/esm/utils').isLoaderWorker()) {
     const userLoaderPaths = getOptionValue('--experimental-loader');
     if (userLoaderPaths.length > 0) {
-      if (!emittedExperimentalWarning) {
-        emitExperimentalWarning('Custom ESM Loaders');
-        emittedExperimentalWarning = true;
+      if (!emittedLoaderFlagWarning) {
+        process.emitWarning(
+          '`--experimental-loader` may be removed in the future; instead use `register()`:\n' +
+          `--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; ${ArrayPrototypeJoin(
+            ArrayPrototypeMap(userLoaderPaths, (loader) => `register(${JSONStringify(encodeURI(loader))}, pathToFileURL("./"))`),
```

Comment:
nit: it makes slightly more to encode the result of `JSON.stringify` rather than JSON-stringify a URL-encoded string. Probably doesn't change anything to the output, but it's easier to follow IMO.
```suggestion
            ArrayPrototypeMap(userLoaderPaths, (loader) => `register(${encodeURI(JSONStringify(loader))}, pathToFileURL("./"))`),
```

Reply:
Also we must use `encodeURIComponent`, we are trying to protect against a loader which would have `?` or `#` in its name, which are two charaacters that `encodeURI` leaves alone for some reason
```suggestion
            ArrayPrototypeMap(userLoaderPaths, (loader) => `register(${encodeURIComponent(JSONStringify(loader))}, pathToFileURL("./"))`),
```

Reply:
> we must use encodeURIComponent

No, that encodes slashes which we don't want. The user needs to see the path.

Reply:
Let's imagine the user is doing

```sh
node --loader './myloader.mjs?key=value#hash' entryPoint.js
```

If we do not escape the `?` and `#`, like that:

```sh
node --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url";register("./myloader.mjs?key=value#hash", pathToFileURL("./"))' entry.js
```

It's going to error out:

```
import { register } from "node:module"; import { pathToFileURL } from "node:url";register("./myloader.mjs
                                                                                          ^^^^^^^^^^^^^^^

SyntaxError: Invalid or unexpected token
```

We must escape `?` and `#`, using `encoreURIComponent` (it also encodes `/`, but not sure why that matters)

Reply:
It's unreadable when the slashes are escaped.

Reply:
That's besides the point, it works. Not escaping `?` and `#` doesn't.

Reply:
This is an error message instructing users what to do. It needs to be both correct and easy to read and understand.

Reply:
I don't really care about readability, but I feel quite strongly that it should be correct. Maybe a compromise would be to check `RegExpPrototypeExec(/[?#]/, specifier) !== null`, and use `encoreURIComponent` only then?

Reply:
We can use `encodeURIComponent` and then replace the escaped slashes with unescaped slashes? Sorry on phone now.

Reply:
Sure, let me push something